### PR TITLE
Add a bunch more guard elimination cases.

### PR DESCRIPTION
**Pull request by resistor**

### CODE REVIEW

File: torch/csrc/jit/passes/guard_elimination.cpp [hunk 1]
```diff
@@ -273,6 +273,21 @@ struct GuardElimination {
     case aten::reciprocal:
     case aten::addcmul:
     case aten::where:
+    case aten::softmax:
+    case aten::log_softmax:
+    case aten::log_sigmoid:
+    case aten::dropout:
+    case aten::matmul:
+    case aten::bmm:
+    case aten::to:
```
Comment (eellison) [inline torch/csrc/jit/passes/guard_elimination.cpp]: I don't think this is right. `checkInputs()` checks if all inputs are summarized, or constant, or a subtype of Number. ScalarTypes are currently represented as an Integer, it is a subtype of Number. That means if you have a variable dtype which is `int` the first 10 iterations, but then `float` the next after, this pass will incorrectly remove the guard on the output being an integer tensor.

Comment (eellison) [reply torch/csrc/jit/passes/guard_elimination.cpp]: I would imagine this is also possible with some of the other cases that are in guard elimination.

Comment (resistor) [reply torch/csrc/jit/passes/guard_elimination.cpp]: I'll remove `aten::to`.  Do you have any others that you think are incorrect?  To me it looks like the rest of them should be OK?

Comment (eellison) [reply torch/csrc/jit/passes/guard_elimination.cpp]: anything that has a `dtype` or `layout` argument, probably. i know at least `aten::rand_like`. one short term fix is to require a constant value if an integer type input the name `dtype` or `layout`. cc @Krovatkin .

File: torch/csrc/jit/passes/guard_elimination.cpp [hunk 2]
```diff
@@ -273,6 +273,20 @@ struct GuardElimination {
...
+    case aten::matmul:
+    case aten::bmm:
+    case aten::_cast_Long:
+    case aten::_cast_Short:
+    case aten::_cast_Int:
+    case aten::_cast_Float:
+    case aten::_cast_Half:
+    case aten::_cast_Double:
+    case aten::contiguous:
+    case aten::select:
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/guard_elimination.cpp]: `select` probably needs to check that `dim` and `index` are both constants

File: torch/csrc/jit/passes/guard_elimination.cpp [hunk 3]
```diff
@@ -273,6 +273,20 @@ struct GuardElimination {
...
+    case aten::dropout:
+    case aten::matmul:
+    case aten::bmm:
+    case aten::_cast_Long:
+    case aten::_cast_Short:
+    case aten::_cast_Int:
+    case aten::_cast_Float:
+    case aten::_cast_Half:
+    case aten::_cast_Double:
+    case aten::contiguous:
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/guard_elimination.cpp]: i can't seem to find a lot of what `aten::contiguous` actually does?

**Review by Krovatkin — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/passes/guard_elimination.cpp (hunk 3 above)
Comment (resistor) [reply torch/csrc/jit/passes/guard_elimination.cpp]: I'm pretty sure it forces the tensor to be contiguous without changing the shape or value.

File: torch/csrc/jit/passes/guard_elimination.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/guard_elimination.cpp]: agreed. i'm going to add `allow_numbers = false` to `checkInputs` or maybe remove `||a subtype of Number` altogether, then the cases like `clamp` can be just implemented with exceptions on specific inputs, but in a short term, `allow_numbers` might be a bit more practical.

**Pull request closed without merging**